go/appencryption/session_cache_test.go

```diff
@@ -0,0 +1,376 @@
+		ttl := time.Millisecond * 100
+
+		// can't use more than 16 sessions here as that is the max drain
+		// for the mango cache implementation
+		totalSessions := 16
+		b := newSessionBucket()
+
+		policy.SessionCacheDuration = ttl
+
+		cache := appencryption.NewSessionCache(b.load, policy)
+		require.NotNil(t, cache)
+
+		defer cache.Close()
+
+		for i := 0; i < totalSessions; i++ {
+			cache.Get(strconv.Itoa(i))
+		}
+
+		// ensure the ttl has elapsed
+		time.Sleep(ttl + time.Millisecond*50)
+
+		expectedCount := 0
+
+		if policy.SessionCacheEngine == "mango" {
+			// mango cache implementation only reaps expired entries following a write, so we'll write a new
+			// cache entry and ensure it's the only one left
+			_, _ = cache.Get("99") // IDs 0-15 were created above
+			expectedCount = 1
+		}
+
+		assert.Eventually(t, func() bool {
+			return cache.Count() == expectedCount
+		}, time.Second*10, time.Millisecond*10)
+	})
+}
+
+func TestSharedSessionCloseOnCacheClose(t *testing.T) {
+	withEachEngine(t, func(t *testing.T, policy *appencryption.CryptoPolicy) {
+		b := newSessionBucket()
+
+		cache := appencryption.NewSessionCache(b.load, policy)
+		require.NotNil(t, cache)
+
+		s, err := cache.Get("my-item")
+		require.NoError(t, err)
+		require.NotNil(t, s)
+
+		s.Close()
+		cache.Close()
+
+		assert.Eventually(t, func() bool {
+			return b.IsClosed(s)
+		}, time.Second*10, time.Millisecond*100)
+	})
+}
+
+func TestSharedSessionCloseOnEviction(t *testing.T) {
+	withEachEngine(t, func(t *testing.T, policy *appencryption.CryptoPolicy) {
+		b := newSessionBucket()
+
+		policy.SessionCacheMaxSize = 1
+
+		cache := appencryption.NewSessionCache(b.load, policy)
+		require.NotNil(t, cache)
+
+		defer cache.Close()
+
+		s1, err := cache.Get("0")
+		require.NoError(t, err)
+		require.NotNil(t, s1)
+		assert.Eventually(t, func() bool {
+			return cache.Count() == 1
+		}, time.Second*10, time.Millisecond*100)
+		s1.Close()
+
+		s2, err := cache.Get("1")
+		require.NoError(t, err)
+		require.NotNil(t, s2)
+
+		assert.Eventually(t, func() bool {
+			return b.IsClosed(s1)
+		}, time.Second*10, time.Millisecond*100)
+
+		assert.False(t, b.IsClosed(s2))
+	})
+}
+
+func TestSharedSessionClose(t *testing.T) {
+	withEachEngine(t, func(t *testing.T, policy *appencryption.CryptoPolicy) {
+		b := newSessionBucket()
+
+		cache := appencryption.NewSessionCache(b.load, policy)
+		require.NotNil(t, cache)
+
+		defer cache.Close()
+
+		for i := 0; i < 2; i++ {
+			s1, err := cache.Get("my-item")
+			require.NoError(t, err)
+			require.NotNil(t, s1)
+			s1.Close()
+
+			assert.Eventually(t, func() bool {
+				return cache.Count() == 1
+			}, time.Second*10, time.Millisecond*100)
+
+			time.Sleep(time.Millisecond * 200)
+
+			// shared sessions arent' actually closed until evicted from the cache
+			assert.False(t, b.IsClosed(s1))
+		}
+	})
+}
+
+func TestSharedSessionNotClosedWhenInUse(t *testing.T) {
```

Comment:
can we rename this similar to `TestSessionCacheGetSessionWhileStillUsedAndNotExpiredShouldNotEvict`

Reply:
actually meant to remove this test as it's functionally equivalent to `TestSharedSessionClose`. will remove